src/lib/components/UrlChooser/index.js

```diff
@@ -0,0 +1,185 @@
+        <div class="lh-enterurl lh-enterurl--switch">
+          <input
+            type="url"
+            class="lh-input"
+            .value="${this.href}"
+            placeholder="Enter a web page URL"
+            pattern="https?://.*"
+            minlength="7"
+            required
+            @keyup="${this.onUrlKeyup}"
+          />
+          <button
+            class="lh-enterurl__close"
+            aria-label="Remove URL"
+            @click=${this.clearInput}
+          ></button>
+        </div>
+        <div class="lh-controls">
+          <button
+            ?disabled=${this.active}
+            class="w-button w-button--secondary"
+            @click=${this.switchUrl}
+          >
+            Switch URL
+          </button>
+          <button
+            ?disabled=${this.active}
+            class="w-button w-button--primary"
+            id="run-lh-button"
+            @click=${this.runAudit}
+          >
+            Run Audit
+          </button>
+        </div>
+      </div>
+    `;
+  }
+
+  updated() {
+    this.urlInput = this.renderRoot.querySelector('input[type="url"]');
+    this.runLighthouseButton = this.renderRoot.querySelector("#run-lh-button");
+  }
+
+  onStateChanged() {
+    const state = store.getState();
+    const urlInputValue = this.urlInput ? this.urlInput.value : null;
+
+    if (this.active) {
+      // do nothing, the URL is being run through Lighthouse
+    } else if (
+      this.href == null &&
+      this.switching &&
+      state.userUrl &&
+      !urlInputValue
+    ) {
+      // if the user has just signed in, the element was in an initial state,
+      // AND the user hasn"t typed anything, reset element with URL
+      this.href = state.userUrl;
+      this.switching = false;
+    } else if (state.userUrl == null && !this.switching) {
+      // if the user has signed out, clear the href and enter switching mode
+      this.href = null;
+      this.switching = true;
+    } else if (state.userUrl && !this.switching) {
+      // in all other cases, only update the URL if ther user isn"t switching
+      this.href = state.userUrl;
+    }
+  }
+
+  runAudit(ev) {
+    if (this.switching) {
+      if (!this.fixUpAndValidateUrl()) {
+        return;
+      }
+
+      // write URL to firestore and local
+      const url = this.urlInput.value;
+      store.setState({userUrl: url});
+      const ref = userRef();
+      if (ref) {
+        const p = ref.set(
+          {
+            userUrl: url,
+            userUrlUpdate: firebase.firestore.FieldValue.serverTimestamp(),
+          },
+          {merge: true},
+        );
+        p.catch((err) => {
+          // TODO: Firestore errors are problematic but we can still run with the new URL.
+          console.warn("could not write URL", err);
+        });
+      }
+
+      this.switching = false;
+    }
+
+    // TODO: actually run Lighthouse. Currently just freezes the page.
```

Comment:
Should this element be running Lighthouse or should it just tell the outside world that it has a URL so a container component knows to then go run Lighthouse.

Reply:
Btw I should mention that for the ProfileSwitcher I used the React notion of Container components (Polymer also often refers to this as Mediator components). https://medium.com/@learnreact/container-components-c0e67432e005

So I was wondering if we'd want to do the same thing with the UrlChooser. It would be a rather dumb component (just moving through its various states depending on if the user has put in a URL or not) and otherwise it would communicate up to its Container when it was time to run Lighthouse.

Reply:
One last discussion point: I think a common practice for Container components is that they are the ones who talk to redux/unistore and handle application state. Their children just handle local state and inform the Container when they've changed. So one question would be if we decide to Containerize this thing, if we would want to move the calls to the store out of this widget and into its Container.

Reply:
On second thought I think I definitely want to pull the actual audit running logic out of this component. It feels like it's overloaded to both have it render the UI and also handle to logic for calling out to Lighthouse.

It also feels like when Lighthouse is running we should reflect that as a state in the store. There are other widgets that may want that info if we decide to break things up into smaller UI chunks.

I think we could put the logic that talks to Lighthouse either into a state action function, or into its own dedicated class. This widget would signal to the Container that it's time to rock and the container would either call the state action or call out to the class object.

Reply:
It now fires `audit` which requests that the controller do something, which could then talk to the state action or class object. For now though I've left that part as a TODO.